056-zk-kraft-migration.md

```diff
@@ -0,0 +1,451 @@
+The FSM is in the `None` state (as per `Kafka.status.migrationState` field).
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: pre-migration` annotation on the `Kafka` custom resource.
+
+**_Actions_**
+
+The operator starts a reconciliation and updates the `Kafka.status` by adding a warning condition about the cluster being already KRaft-based, so no migration is possible.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM stays in the `None` state.
+The `Kafka.status` keeps the warning condition, on each subsequent reconciliation, until the user re-applies the `strimzi.io/kraft: enabled` annotation.
+
+#### Initializing migration process
+
+**_Preconditions_**
+
+The Apache Kafka cluster is running, meets the prerequisites described above, and is in a functional state.
+The `strimzi.io/kraft: disabled` annotation is already applied on the `Kafka` custom resource or missing, denoting a ZooKeeper-based cluster.
+The FSM is in the `NoMigration` (or null) state (as per `Kafka.status.migrationState` field).
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: pre-migration` annotation on the `Kafka` custom resource to make the operator aware of the migration intention.
+
+**_Actions_**
+
+The operator starts a reconciliation but there are no actual actions to be executed.
+The `KRaftMigrationManager` component just prepares itself, with an internal status change, to be ready for the migration process.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves to the `WaitControllersInMigration` state where the operator expects the user to create the KRaft controllers pool.
+
+#### Phase 1: Provisioning the KRaft controller quorum
+
+At the beginning of this phase, it's still possible to back out of the migration.
+
+**_Preconditions_**
+
+The FSM is in the `WaitControllersInMigration` state (as per `Kafka.status.migrationState` field).
+
+**_Trigger_**
+
+The user sets the `strimzi.io/kraft: disabled` annotation on the `Kafka` custom resource.
+
+**_Actions_**
+
+The operator starts a reconciliation but there are no actual actions to be executed.
+The `KRaftMigrationManager` component just comes back to idle, with an internal status change.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to the `NoMigration` state (or null) actually exiting from the migration process and the operator upgrades `Kafka.status.migrationState` field accordingly.
+
+---
+
+If not rolling back, the user can move forward by deploying the KRaft controllers pool.
+
+**_Preconditions_**
+
+The FSM is in the `WaitControllersInMigration` state (as per `Kafka.status.migrationState` field).
+
+**_Trigger_**
+
+The user creates the `KafkaNodePool` custom resource containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
```

Comment:
What if the node pool already exists when the user sets the `pre-migration` annotation?

Reply:
Isn't it the case of the issue I opened but you said it's not how it should be used? So use will just screw up because what happens is described here: https://github.com/strimzi/strimzi-kafka-operator/issues/9108

Reply:
This goes more to the automation point. I expect many users when presented with it like this to take the shortcut and do both steps in one.

Reply:
Actually it should be solved by https://github.com/strimzi/proposals/pull/93